### PR TITLE
Update ToLength operation to conform ES6 spec

### PR DESCRIPTION
**Pull request by rerobika**
JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/lit/lit-globals.h [hunk 1]
```diff
@@ -78,6 +78,22 @@ typedef uint16_t ecma_char_t;
...
+ * Description of a property index
+ *
+ * For ES6+ the maximum valid property index is 2**53 - 1
+ */
+typedef uint64_t ecma_prop_index_t;
+#else /* !ENABLED (JERRY_ESNEXT) */
+/**
+ * Description of a property index
+ *
+ * For ES6+ the maximum valid property index is 2**32 - 1
```
Comment (ossy-szeged) [inline jerry-core/lit/lit-globals.h:92]: Typo: ES5.1

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 2]
```diff
@@ -569,13 +593,14 @@ ecma_builtin_array_prototype_object_reverse (ecma_value_t this_arg, /**< this ar
...
+  for (ecma_length_t lower = 0; lower < middle; lower++)
   {
-    uint32_t upper = len - lower - 1;
+    ecma_length_t upper = len - lower - 1;
     ecma_value_t ret_value = ECMA_VALUE_ERROR;
 
-    ecma_string_t *lower_str_p = ecma_new_ecma_string_from_uint32 (lower);
-    ecma_string_t *upper_str_p = ecma_new_ecma_string_from_uint32 (upper);
+    ecma_string_t *lower_str_p = ecma_new_ecma_string_from_number ((ecma_number_t) lower);
+    ecma_string_t *upper_str_p = ecma_new_ecma_string_from_number ((ecma_number_t) upper);
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c:603]: How about `ecma_new_ecma_string_from_length`?

File: jerry-core/ecma/operations/ecma-objects.c [hunk 3]
```diff
@@ -680,40 +680,31 @@ ecma_op_object_find_own (ecma_value_t base_value, /**< base value */
...
 
-  ecma_string_t *index_str_p = ecma_new_non_direct_string_from_uint32 (index);
-  ecma_value_t ret_value = ecma_op_object_find (object_p, index_str_p);
-  ecma_deref_ecma_string (index_str_p);
-
-  return ret_value;
-} /* ecma_op_object_find_by_uint32_index */
+  ecma_string_t *index_str_p;
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_UNLIKELY (index > UINT32_MAX))
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-objects.c:693]: Are these functions that much of a bottleneck to warrant manually inlining these? Just curious, to me it seems the performance cost of calling for example `ecma_new_ecma_string_from_length` would be negligible compared to the find operation itself.

File: jerry-core/ecma/operations/ecma-array-object.c [hunk 4]
```diff
@@ -698,7 +698,7 @@ ecma_op_array_species_create (ecma_object_t *original_array_p, /**< The object f
 
   if (ecma_is_value_undefined (constructor))
   {
-    ecma_value_t length_val = ecma_make_uint32_value (length);
+    ecma_value_t length_val = ecma_make_number_value ((ecma_number_t) length);
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-array-object.c:701]: `ecma_make_length_value`?

File: jerry-core/ecma/operations/ecma-array-object.c [hunk 5]
```diff
@@ -711,7 +711,7 @@ ecma_op_array_species_create (ecma_object_t *original_array_p, /**< The object f
     return ecma_raise_type_error (ECMA_ERR_MSG ("Invalid species constructor"));
   }
 
-  ecma_value_t len_val = ecma_make_uint32_value (length);
+  ecma_value_t len_val = ecma_make_number_value ((ecma_number_t) length);
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-array-object.c:714]: `ecma_make_length_value`?

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 6]
```diff
@@ -549,7 +550,30 @@ ecma_new_non_direct_string_from_uint32 (uint32_t uint32_number) /**< uint32 valu
...
+{
+  if (JERRY_LIKELY (number <= ECMA_DIRECT_STRING_MAX_IMM))
+  {
+    return (ecma_string_t *) ECMA_CREATE_DIRECT_STRING (ECMA_DIRECT_STRING_UINT, (uintptr_t) number);
+  }
+
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_UNLIKELY (number > UINT32_MAX))
+  {
+    return ecma_new_ecma_string_from_number ((ecma_number_t) number);
```
Comment (dbatyai) [inline jerry-core/ecma/base/ecma-helpers-string.c:568]: `JERRY_ASSERT (number <= ECMA_NUMBER_MAX_SAFE_INTEGER);` would be useful here.

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 7]
```diff
@@ -530,6 +530,22 @@ ecma_is_number_equal_to_positive_zero (ecma_number_t ecma_number) /**< number */
...
+ */
+ecma_value_t
+ecma_make_length_value (ecma_length_t number) /**< number to be encoded */
+{
+  if (number <= ECMA_INTEGER_NUMBER_MAX)
+  {
+    return ecma_make_integer_value ((ecma_integer_value_t) number);
+  }
+
+  return ecma_create_float_number ((ecma_number_t) number);
```
Comment (dbatyai) [inline jerry-core/ecma/base/ecma-helpers-value.c:546]: `JERRY_ASSERT (number <= ECMA_NUMBER_MAX_SAFE_INTEGER);` here as well.

**Comment by rerobika** [comment]
@dbatyai Thanks for the review, I've updated the PR.

**Comment by ossy-szeged** [comment]
LGTM (informal)

**Pull request merged**